reactor-core/src/test/java/reactor/core/publisher/FluxSwitchOnFirstTest.java

```diff
@@ -0,0 +1,677 @@
+                    .expectError()
+                    .verifyThenAssertThat()
+                    .hasDroppedErrors(3)
+                    .tookLessThan(Duration.ofSeconds(10));
+
+        publisher.assertWasRequested();
+        publisher.assertNoRequestOverflow();
+    }
+
+    //    @Test
+    //    public void shouldNotFailOnIncorrePu
+
+    @Test
+    public void shouldBeAbleToAccessUpstreamContext() {
+        TestPublisher<Long> publisher = TestPublisher.createCold();
+
+        Flux<String> switchTransformed = publisher.flux()
+                                                  .switchOnFirst(
+                                                      (first, innerFlux) -> innerFlux.map(String::valueOf)
+                                                                                     .subscriberContext(Context.of("a", "b"))
+                                                  )
+                                                  .subscriberContext(Context.of("a", "c"))
+                                                  .subscriberContext(Context.of("c", "d"));
+
+        publisher.next(1L);
+
+        StepVerifier.create(switchTransformed, 0)
+                    .thenRequest(1)
+                    .expectNext("1")
+                    .thenRequest(1)
+                    .then(() -> publisher.next(2L))
+                    .expectNext("2")
+                    .expectAccessibleContext()
+                    .contains("a", "b")
+                    .contains("c", "d")
+                    .then()
+                    .then(publisher::complete)
+                    .expectComplete()
+                    .verify(Duration.ofSeconds(10));
+
+        publisher.assertWasRequested();
+        publisher.assertNoRequestOverflow();
+    }
+
+    @Test
+    public void shouldNotHangWhenOneElementUpstream() {
+        TestPublisher<Long> publisher = TestPublisher.createCold();
+
+        Flux<String> switchTransformed = publisher.flux()
+                                                  .switchOnFirst((first, innerFlux) ->
+                                                      innerFlux.map(String::valueOf)
+                                                               .subscriberContext(Context.of("a", "b"))
+                                                  )
+                                                  .subscriberContext(Context.of("a", "c"))
+                                                  .subscriberContext(Context.of("c", "d"));
+
+        publisher.next(1L);
+        publisher.complete();
+
+        StepVerifier.create(switchTransformed, 0)
+                    .thenRequest(1)
+                    .expectNext("1")
+                    .expectComplete()
+                    .verify(Duration.ofSeconds(10));
+
+        publisher.assertWasRequested();
+        publisher.assertNoRequestOverflow();
+    }
+
+    @Test
+    public void backpressureTest() {
+        TestPublisher<Long> publisher = TestPublisher.createCold();
```

Comment:
it should be documented in the API that the source is expected to be cold. Also, in order to make it evident that the source isn't subscribed to twice, an explicit test like below would make sense:

```java
    @Test
    public void sourceSubscribedOnce() {
        AtomicInteger subCount = new AtomicInteger();
        Flux<Integer> source = Flux.range(1, 10)
                                   .hide()
                                   .doOnSubscribe(subscription -> subCount.incrementAndGet());

        StepVerifier.create(source.switchOnFirst((s, f) -> f.filter(v -> v % 2 == s.get())))
                    .expectNext(1, 3, 5, 7, 9)
                    .verifyComplete();
        
        assertThat(subCount).hasValue(1);
    }
```

Reply:
what do you mean by `cold`? The source could be `hot` as well.

Will add such a test. In case you point to the fact that switch transform does not cause double subscription, that it would make sence

Reply:
yeah, plus a bunch of these tests use `TestPublisher.createCold()` and don't work with the `TestPublisher.create()` (which can be viewed as hot, it forwards the same data to all subscribers)